src/braket/aws/aws_emulator_helpers.py

```diff
@@ -0,0 +1,200 @@
+
+
+@singledispatch
+def create_gate_connectivity_criterion(
+    properties: DeviceCapabilities, connectivity_graph: DiGraph
+) -> GateConnectivityCriterion:
+    raise NotImplementedError
+
+
+@create_gate_connectivity_criterion.register(IqmDeviceCapabilities)
+@create_gate_connectivity_criterion.register(RigettiDeviceCapabilities)
+def _(
+    properties: RigettiDeviceCapabilities, connectivity_graph: DiGraph
+) -> GateConnectivityCriterion:
+    """
+    Both IQM and Rigetti have undirected connectivity graphs; Rigetti device capabilities
+    provide back edges, but the calibration data only provides edges in one direction.
+    Additionally, IQM does not provide back edges in its connectivity_graph (nor is this
+    resolved manually by AwsDevice at the moment).
+    """
+    gate_connectivity_graph = connectivity_graph.copy()
+    edge_properties = properties.standardized.twoQubitProperties
+    for u, v in gate_connectivity_graph.edges:
+        edge_key = "-".join([str(qubit) for qubit in (u, v)])
+        edge_property = edge_properties.get(edge_key)
+
+        # Check that the QHP provided calibration data for this edge.
+        if not edge_property:
+            gate_connectivity_graph[u][v]["supported_gates"] = set()
+            continue
+        edge_supported_gates = get_qpu_gate_translation(
+            properties, [property.gateName for property in edge_property.twoQubitGateFidelity]
+        )
+        gate_connectivity_graph[u][v]["supported_gates"] = set(edge_supported_gates)
+
+    # Add the reversed edge to ensure gates can be applied
+    # in both directions for a given qubit pair.
+    for u, v in gate_connectivity_graph.edges:
+        if (v, u) not in gate_connectivity_graph.edges or gate_connectivity_graph[v][u].get(
+            "supported_gates"
+        ) in [None, set()]:
+            gate_connectivity_graph.add_edge(
+                v, u, supported_gates=set(gate_connectivity_graph[u][v]["supported_gates"])
+            )
+
+    return GateConnectivityCriterion(gate_connectivity_graph)
+
+
+@create_gate_connectivity_criterion.register(IonqDeviceCapabilities)
+def _(properties: IonqDeviceCapabilities, connectivity_graph: DiGraph) -> GateConnectivityCriterion:
+    """
+    Qubits in IonQ's trapped ion devices are all fully connected with identical
+    gate-pair capabilities. IonQ does not expliclty provide a set of edges for
+    gate connectivity between qubit pairs in their trapped ion QPUs.
+    We extrapolate gate connectivity across all possible qubit edge pairs.
+    """
+    gate_connectivity_graph = connectivity_graph.copy()
+    native_gates = get_qpu_gate_translation(properties, properties.paradigm.nativeGateSet)
+
+    for edge in gate_connectivity_graph.edges:
+        gate_connectivity_graph[edge[0]][edge[1]]["supported_gates"] = set(native_gates)
+
+    return GateConnectivityCriterion(gate_connectivity_graph)
+
+
+def get_qpu_gate_translation(
```

Comment:
Are the names not already standardized across QPUs? This shouldn't be necessary.

Reply:
There are a few places where translation to a Braket gate name was necessary, i.e. gate names in Rigetti calibration data ("CPHASE" is used instead of the standard "cphaseshift") or in IonQs native gate set (translating "GPI"/"GPI2" to "GPi"/"GPi2"). I felt this was a little cleaner than doing translations in each place required.